{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting transformers\n",
      "  Downloading transformers-4.32.1-py3-none-any.whl (7.5 MB)\n",
      "     ---------------------------------------- 7.5/7.5 MB 2.1 MB/s eta 0:00:00\n",
      "Requirement already satisfied: packaging>=20.0 in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from transformers) (23.0)\n",
      "Requirement already satisfied: tqdm>=4.27 in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from transformers) (4.64.1)\n",
      "Requirement already satisfied: requests in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from transformers) (2.28.2)\n",
      "Collecting tokenizers!=0.11.3,<0.14,>=0.11.1\n",
      "  Downloading tokenizers-0.13.3-cp39-cp39-win_amd64.whl (3.5 MB)\n",
      "     ---------------------------------------- 3.5/3.5 MB 2.8 MB/s eta 0:00:00\n",
      "Collecting huggingface-hub<1.0,>=0.15.1\n",
      "  Downloading huggingface_hub-0.16.4-py3-none-any.whl (268 kB)\n",
      "     -------------------------------------- 268.8/268.8 kB 8.3 MB/s eta 0:00:00\n",
      "Requirement already satisfied: numpy>=1.17 in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from transformers) (1.23.5)\n",
      "Collecting safetensors>=0.3.1\n",
      "  Downloading safetensors-0.3.3-cp39-cp39-win_amd64.whl (266 kB)\n",
      "     ------------------------------------- 266.4/266.4 kB 16.0 MB/s eta 0:00:00\n",
      "Requirement already satisfied: pyyaml>=5.1 in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from transformers) (6.0)\n",
      "Requirement already satisfied: filelock in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from transformers) (3.10.0)\n",
      "Collecting regex!=2019.12.17\n",
      "  Downloading regex-2023.8.8-cp39-cp39-win_amd64.whl (268 kB)\n",
      "     ------------------------------------- 268.4/268.4 kB 16.1 MB/s eta 0:00:00\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from huggingface-hub<1.0,>=0.15.1->transformers) (4.5.0)\n",
      "Requirement already satisfied: fsspec in c:\\users\\ilkhom-pc\\appdata\\roaming\\python\\python39\\site-packages (from huggingface-hub<1.0,>=0.15.1->transformers) (2023.1.0)\n",
      "Requirement already satisfied: colorama in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from tqdm>=4.27->transformers) (0.4.6)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from requests->transformers) (1.26.15)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from requests->transformers) (2022.12.7)\n",
      "Requirement already satisfied: idna<4,>=2.5 in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from requests->transformers) (3.4)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages (from requests->transformers) (2.1.1)\n",
      "Installing collected packages: tokenizers, safetensors, regex, huggingface-hub, transformers\n",
      "Successfully installed huggingface-hub-0.16.4 regex-2023.8.8 safetensors-0.3.3 tokenizers-0.13.3 transformers-4.32.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: Ignoring invalid distribution -pencv-python (c:\\users\\ilkhom-pc\\appdata\\roaming\\python\\python39\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (c:\\users\\ilkhom-pc\\appdata\\roaming\\python\\python39\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (c:\\users\\ilkhom-pc\\appdata\\roaming\\python\\python39\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (c:\\users\\ilkhom-pc\\appdata\\roaming\\python\\python39\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (c:\\users\\ilkhom-pc\\appdata\\roaming\\python\\python39\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (c:\\users\\ilkhom-pc\\appdata\\roaming\\python\\python39\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n"
     ]
    }
   ],
   "source": [
    "!pip install transformers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Package                      Version"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: Ignoring invalid distribution -pencv-python (c:\\users\\ilkhom-pc\\appdata\\roaming\\python\\python39\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (c:\\users\\ilkhom-pc\\appdata\\roaming\\python\\python39\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (c:\\users\\ilkhom-pc\\appdata\\roaming\\python\\python39\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python (c:\\users\\ilkhom-pc\\appdata\\roaming\\python\\python39\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -otobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -rotobuf (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -pencv-python-headless (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ensorflow (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -atplotlib (d:\\programmfiles\\miniconda\\envs\\cv\\lib\\site-packages)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "---------------------------- --------------------\n",
      "absl-py                      1.4.0\n",
      "aiohttp                      3.8.3\n",
      "aiosignal                    1.3.1\n",
      "albumentations               1.3.0\n",
      "appdirs                      1.4.4\n",
      "asttokens                    2.2.1\n",
      "astunparse                   1.6.3\n",
      "async-timeout                4.0.2\n",
      "asyncio                      3.4.3\n",
      "attrs                        22.2.0\n",
      "audioread                    3.0.0\n",
      "backcall                     0.2.0\n",
      "beautifulsoup4               4.12.0\n",
      "cachetools                   5.3.0\n",
      "certifi                      2022.12.7\n",
      "cffi                         1.15.1\n",
      "charset-normalizer           2.1.1\n",
      "click                        8.1.3\n",
      "cloudpickle                  2.2.1\n",
      "colorama                     0.4.6\n",
      "comm                         0.1.2\n",
      "contourpy                    1.0.7\n",
      "cycler                       0.11.0\n",
      "Cython                       0.29.33\n",
      "dask                         2023.4.0\n",
      "dask-glm                     0.2.0\n",
      "dask-ml                      2023.3.24\n",
      "debugpy                      1.6.6\n",
      "decorator                    5.1.1\n",
      "deepface                     0.0.79\n",
      "distributed                  2023.4.0\n",
      "dm-tree                      0.1.8\n",
      "docker-pycreds               0.4.0\n",
      "etils                        1.1.1\n",
      "executing                    1.2.0\n",
      "filelock                     3.10.0\n",
      "fire                         0.5.0\n",
      "Flask                        2.2.3\n",
      "flatbuffers                  23.3.3\n",
      "fonttools                    4.38.0\n",
      "frozenlist                   1.3.3\n",
      "fsspec                       2023.1.0\n",
      "gast                         0.4.0\n",
      "gdown                        4.6.4\n",
      "gevent                       22.10.2\n",
      "gitdb                        4.0.10\n",
      "GitPython                    3.1.31\n",
      "google-auth                  2.16.2\n",
      "google-auth-oauthlib         0.4.6\n",
      "google-pasta                 0.2.0\n",
      "googleapis-common-protos     1.59.0\n",
      "greenlet                     2.0.2\n",
      "grpcio                       1.51.3\n",
      "gunicorn                     20.1.0\n",
      "h5py                         3.8.0\n",
      "idna                         3.4\n",
      "imageio                      2.25.0\n",
      "imbalanced-learn             0.10.1\n",
      "imblearn                     0.0\n",
      "imgaug                       0.4.0\n",
      "importlib-metadata           6.1.0\n",
      "importlib-resources          5.12.0\n",
      "imutils                      0.5.4\n",
      "ipykernel                    6.21.1\n",
      "ipython                      8.9.0\n",
      "ipython-genutils             0.2.0\n",
      "ipywidgets                   8.0.4\n",
      "itsdangerous                 2.1.2\n",
      "jedi                         0.18.2\n",
      "Jinja2                       3.1.2\n",
      "joblib                       1.2.0\n",
      "jsonpickle                   3.0.1\n",
      "jupyter_client               8.0.2\n",
      "jupyter_core                 5.2.0\n",
      "jupyterlab-widgets           3.0.5\n",
      "keras                        2.11.0\n",
      "Keras-Applications           1.0.8\n",
      "keras-nightly                2.5.0.dev2021032900\n",
      "Keras-Preprocessing          1.1.2\n",
      "keras-tuner                  1.3.5\n",
      "keras-vggface                0.6\n",
      "kiwisolver                   1.4.4\n",
      "kt-legacy                    1.0.5\n",
      "lazy_loader                  0.2\n",
      "libclang                     15.0.6.1\n",
      "librosa                      0.10.0.post2\n",
      "lightning-utilities          0.8.0\n",
      "llvmlite                     0.39.1\n",
      "locket                       1.0.0\n",
      "lxml                         4.9.2\n",
      "Markdown                     3.4.1\n",
      "MarkupSafe                   2.1.2\n",
      "matplotlib                   3.7.1\n",
      "matplotlib-inline            0.1.6\n",
      "modelhub-client              1.1.0\n",
      "mpmath                       1.3.0\n",
      "msgpack                      1.0.5\n",
      "mtcnn                        0.1.1\n",
      "multidict                    6.0.4\n",
      "multipledispatch             0.6.0\n",
      "nest-asyncio                 1.5.6\n",
      "networkx                     3.0\n",
      "numba                        0.56.4\n",
      "numpy                        1.23.5\n",
      "oauthlib                     3.2.2\n",
      "opencv-python                4.8.0.74\n",
      "opt-einsum                   3.3.0\n",
      "packaging                    23.0\n",
      "pandas                       1.5.3\n",
      "pandas-datareader            0.10.0\n",
      "parso                        0.8.3\n",
      "partd                        1.4.0\n",
      "pathtools                    0.1.2\n",
      "patsy                        0.5.3\n",
      "pickleshare                  0.7.5\n",
      "Pillow                       9.4.0\n",
      "pip                          22.3.1\n",
      "platformdirs                 3.0.0\n",
      "plotly                       5.14.1\n",
      "pooch                        1.6.0\n",
      "promise                      2.3\n",
      "prompt-toolkit               3.0.36\n",
      "protobuf                     3.19.6\n",
      "psutil                       5.9.4\n",
      "pure-eval                    0.2.2\n",
      "pyasn1                       0.4.8\n",
      "pyasn1-modules               0.2.8\n",
      "pycocotools                  2.0.6\n",
      "pycparser                    2.21\n",
      "Pygments                     2.14.0\n",
      "pyparsing                    3.0.9\n",
      "PySocks                      1.7.1\n",
      "python-dateutil              2.8.2\n",
      "python-version               0.0.2\n",
      "PyTurboJPEG                  1.7.0\n",
      "pytz                         2022.7.1\n",
      "PyWavelets                   1.4.1\n",
      "pywin32                      305\n",
      "PyYAML                       6.0\n",
      "pyzmq                        25.0.0\n",
      "qudida                       0.0.4\n",
      "randomsearch                 0.0.1\n",
      "requests                     2.28.2\n",
      "requests-oauthlib            1.3.1\n",
      "retina-face                  0.0.13\n",
      "rsa                          4.9\n",
      "scikit-image                 0.19.3\n",
      "scikit-learn                 1.2.1\n",
      "scikit-surprise              1.1.3\n",
      "scipy                        1.10.1\n",
      "seaborn                      0.12.2\n",
      "sentry-sdk                   1.17.0\n",
      "setproctitle                 1.3.2\n",
      "setuptools                   68.0.0\n",
      "shapely                      2.0.1\n",
      "six                          1.16.0\n",
      "sklearn                      0.0.post1\n",
      "smmap                        5.0.0\n",
      "sortedcontainers             2.4.0\n",
      "soundfile                    0.12.1\n",
      "soupsieve                    2.4\n",
      "soxr                         0.3.5\n",
      "stack-data                   0.6.2\n",
      "statsmodels                  0.13.5\n",
      "surprise                     0.1\n",
      "sympy                        1.11.1\n",
      "tblib                        1.7.0\n",
      "tenacity                     8.2.2\n",
      "tensorboard                  2.11.2\n",
      "tensorboard-data-server      0.6.1\n",
      "tensorboard-plugin-wit       1.8.1\n",
      "tensorflow                   2.11.1\n",
      "tensorflow-datasets          4.8.3\n",
      "tensorflow-estimator         2.11.0\n",
      "tensorflow-intel             2.11.1\n",
      "tensorflow-io-gcs-filesystem 0.31.0\n",
      "tensorflow-metadata          1.12.0\n",
      "termcolor                    2.2.0\n",
      "tf-estimator-nightly         2.8.0.dev2021122109\n",
      "thop                         0.1.1.post2209072238\n",
      "threadpoolctl                3.1.0\n",
      "tifffile                     2023.2.3\n",
      "toml                         0.10.2\n",
      "toolz                        0.12.0\n",
      "torch                        1.10.1+cu113\n",
      "torchaudio                   0.10.1+cu113\n",
      "torchmetrics                 0.11.4\n",
      "torchvision                  0.11.2+cu113\n",
      "tornado                      6.2\n",
      "tqdm                         4.64.1\n",
      "traitlets                    5.9.0\n",
      "typeguard                    2.13.3\n",
      "typing_extensions            4.5.0\n",
      "ujson                        5.7.0\n",
      "urllib3                      1.26.15\n",
      "val                          0.6\n",
      "wandb                        0.14.0\n",
      "wcwidth                      0.2.6\n",
      "Werkzeug                     2.2.3\n",
      "wheel                        0.40.0\n",
      "widgetsnbextension           4.0.5\n",
      "wincertstore                 0.2\n",
      "wrapt                        1.15.0\n",
      "yarl                         1.8.2\n",
      "zict                         3.0.0\n",
      "zipp                         3.15.0\n",
      "zope.event                   4.6\n",
      "zope.interface               5.5.2\n"
     ]
    }
   ],
   "source": [
    "!pip list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from transformers import BertTokenizer, TFBertForSequenceClassification\n",
    "from transformers import InputExample, InputFeatures\n",
    "import pandas as pd\n",
    "import tensorflow as tf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_data = pd.read_csv('train.csv')\n",
    "test_data = pd.read_csv('test.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "tokenizer = BertTokenizer.from_pretrained(\"bert-base-multilingual-cased\")\n",
    "\n",
    "def convert_data_to_examples(train, DATA_COLUMN, LABEL_COLUMN): \n",
    "    return train.apply(lambda x: InputExample(guid=None, text_a = x[DATA_COLUMN], text_b = None, label = x[LABEL_COLUMN]), axis = 1)\n",
    "\n",
    "def convert_data_to_examples_without_labels(test, DATA_COLUMN): \n",
    "    return test.apply(lambda x: InputExample(guid=None, text_a = x[DATA_COLUMN], text_b = None, label = 0), axis = 1)\n",
    "\n",
    "def convert_examples_to_tf_dataset(examples, tokenizer):\n",
    "    features = []\n",
    "    for e in examples:\n",
    "        input_dict = tokenizer.encode_plus(e.text_a, add_special_tokens=True, max_length=128, return_token_type_ids=True, return_attention_mask=True, pad_to_max_length=True, truncation=True)\n",
    "        input_ids, token_type_ids, attention_mask = (input_dict[\"input_ids\"], input_dict[\"token_type_ids\"], input_dict['attention_mask'])\n",
    "        features.append(InputFeatures(input_ids=input_ids, attention_mask=attention_mask, token_type_ids=token_type_ids, label=e.label))\n",
    "\n",
    "    def gen():\n",
    "        for f in features:\n",
    "            yield ({\"input_ids\": f.input_ids, \"attention_mask\": f.attention_mask, \"token_type_ids\": f.token_type_ids}, f.label)\n",
    "\n",
    "    return tf.data.Dataset.from_generator(gen, ({\"input_ids\": tf.int32, \"attention_mask\": tf.int32, \"token_type_ids\": tf.int32}, tf.int64))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\transformers\\tokenization_utils_base.py:2418: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "DATA_COLUMN = 'Text'\n",
    "LABEL_COLUMN = 'Sentiment'\n",
    "label_map = {'Negative': 0, 'Neutral': 1, 'Positive': 2, 'Extremely Positive': 3, 'Extremely Negative': 4}\n",
    "\n",
    "# Drop rows where the text column has NaN values\n",
    "train_data = train_data.dropna(subset=[DATA_COLUMN])\n",
    "train_data[DATA_COLUMN] = train_data[DATA_COLUMN].fillna(\"\")\n",
    "\n",
    "# Proceed with your existing code...\n",
    "train_data[LABEL_COLUMN] = train_data[LABEL_COLUMN].map(label_map)\n",
    "train_InputExamples = convert_data_to_examples(train_data, DATA_COLUMN, LABEL_COLUMN)\n",
    "train_data_tf = convert_examples_to_tf_dataset(list(train_InputExamples), tokenizer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "90a518e2035a42e5850821875413d429",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading model.safetensors:   0%|          | 0.00/714M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "All PyTorch model weights were used when initializing TFBertForSequenceClassification.\n",
      "\n",
      "Some weights or buffers of the TF 2.0 model TFBertForSequenceClassification were not initialized from the PyTorch model and are newly initialized: ['classifier.weight', 'classifier.bias']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/3\n"
     ]
    },
    {
     "ename": "OperatorNotAllowedInGraphError",
     "evalue": "in user code:\n\n    File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\keras\\engine\\training.py\", line 1249, in train_function  *\n        return step_function(self, iterator)\n    File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\keras\\engine\\training.py\", line 1233, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\keras\\engine\\training.py\", line 1222, in run_step  **\n        outputs = model.train_step(data)\n    File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\transformers\\modeling_tf_utils.py\", line 1637, in train_step\n        y_pred = self(x, training=True)\n    File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 70, in error_handler\n        raise e.with_traceback(filtered_tb) from None\n\n    OperatorNotAllowedInGraphError: Exception encountered when calling layer 'tf_bert_for_sequence_classification' (type TFBertForSequenceClassification).\n    \n    in user code:\n    \n        File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\transformers\\modeling_tf_utils.py\", line 1557, in run_call_with_unpacked_inputs  *\n            return func(self, **unpacked_inputs)\n        File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\transformers\\models\\bert\\modeling_tf_bert.py\", line 1569, in call  *\n            outputs = self.bert(\n        File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 70, in error_handler  **\n            raise e.with_traceback(filtered_tb) from None\n    \n        OperatorNotAllowedInGraphError: Exception encountered when calling layer 'bert' (type TFBertMainLayer).\n        \n        in user code:\n        \n            File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\transformers\\modeling_tf_utils.py\", line 1557, in run_call_with_unpacked_inputs  *\n                return func(self, **unpacked_inputs)\n            File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\transformers\\models\\bert\\modeling_tf_bert.py\", line 766, in call  *\n                batch_size, seq_length = input_shape\n        \n            OperatorNotAllowedInGraphError: Iterating over a symbolic `tf.Tensor` is not allowed: AutoGraph did convert this function. This might indicate you are trying to use an unsupported feature.\n        \n        \n        Call arguments received by layer 'bert' (type TFBertMainLayer):\n          • self=tf.Tensor(shape=<unknown>, dtype=int32)\n          • input_ids=None\n          • attention_mask=tf.Tensor(shape=<unknown>, dtype=int32)\n          • token_type_ids=tf.Tensor(shape=<unknown>, dtype=int32)\n          • position_ids=None\n          • head_mask=None\n          • inputs_embeds=None\n          • encoder_hidden_states=None\n          • encoder_attention_mask=None\n          • past_key_values=None\n          • use_cache=None\n          • output_attentions=False\n          • output_hidden_states=False\n          • return_dict=True\n          • training=True\n    \n    \n    Call arguments received by layer 'tf_bert_for_sequence_classification' (type TFBertForSequenceClassification):\n      • self={'input_ids': 'tf.Tensor(shape=<unknown>, dtype=int32)', 'attention_mask': 'tf.Tensor(shape=<unknown>, dtype=int32)', 'token_type_ids': 'tf.Tensor(shape=<unknown>, dtype=int32)'}\n      • input_ids=None\n      • attention_mask=None\n      • token_type_ids=None\n      • position_ids=None\n      • head_mask=None\n      • inputs_embeds=None\n      • output_attentions=None\n      • output_hidden_states=None\n      • return_dict=None\n      • labels=None\n      • training=True\n",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mOperatorNotAllowedInGraphError\u001b[0m            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[15], line 4\u001b[0m\n\u001b[0;32m      2\u001b[0m optimizer \u001b[39m=\u001b[39m tf\u001b[39m.\u001b[39mkeras\u001b[39m.\u001b[39moptimizers\u001b[39m.\u001b[39mAdam(learning_rate\u001b[39m=\u001b[39m\u001b[39m5e-5\u001b[39m)\n\u001b[0;32m      3\u001b[0m model\u001b[39m.\u001b[39mcompile(optimizer\u001b[39m=\u001b[39moptimizer, loss\u001b[39m=\u001b[39mmodel\u001b[39m.\u001b[39mcompute_loss, metrics\u001b[39m=\u001b[39m[\u001b[39m\"\u001b[39m\u001b[39maccuracy\u001b[39m\u001b[39m\"\u001b[39m])\n\u001b[1;32m----> 4\u001b[0m model\u001b[39m.\u001b[39;49mfit(train_data_tf\u001b[39m.\u001b[39;49mshuffle(\u001b[39m1000\u001b[39;49m)\u001b[39m.\u001b[39;49mbatch(\u001b[39m16\u001b[39;49m), epochs\u001b[39m=\u001b[39;49m\u001b[39m3\u001b[39;49m, batch_size\u001b[39m=\u001b[39;49m\u001b[39m16\u001b[39;49m)\n\u001b[0;32m      6\u001b[0m validation_InputExamples \u001b[39m=\u001b[39m convert_data_to_examples_without_labels(test_data, DATA_COLUMN)\n\u001b[0;32m      7\u001b[0m test_data_tf \u001b[39m=\u001b[39m convert_examples_to_tf_dataset(\u001b[39mlist\u001b[39m(validation_InputExamples), tokenizer)\n",
      "File \u001b[1;32md:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\keras\\utils\\traceback_utils.py:70\u001b[0m, in \u001b[0;36mfilter_traceback.<locals>.error_handler\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     67\u001b[0m     filtered_tb \u001b[39m=\u001b[39m _process_traceback_frames(e\u001b[39m.\u001b[39m__traceback__)\n\u001b[0;32m     68\u001b[0m     \u001b[39m# To get the full stack trace, call:\u001b[39;00m\n\u001b[0;32m     69\u001b[0m     \u001b[39m# `tf.debugging.disable_traceback_filtering()`\u001b[39;00m\n\u001b[1;32m---> 70\u001b[0m     \u001b[39mraise\u001b[39;00m e\u001b[39m.\u001b[39mwith_traceback(filtered_tb) \u001b[39mfrom\u001b[39;00m \u001b[39mNone\u001b[39m\n\u001b[0;32m     71\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m     72\u001b[0m     \u001b[39mdel\u001b[39;00m filtered_tb\n",
      "File \u001b[1;32md:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\tensorflow\\python\\framework\\func_graph.py:1269\u001b[0m, in \u001b[0;36mfunc_graph_from_py_func.<locals>.autograph_handler\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m   1267\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m e:  \u001b[39m# pylint:disable=broad-except\u001b[39;00m\n\u001b[0;32m   1268\u001b[0m   \u001b[39mif\u001b[39;00m \u001b[39mhasattr\u001b[39m(e, \u001b[39m\"\u001b[39m\u001b[39mag_error_metadata\u001b[39m\u001b[39m\"\u001b[39m):\n\u001b[1;32m-> 1269\u001b[0m     \u001b[39mraise\u001b[39;00m e\u001b[39m.\u001b[39mag_error_metadata\u001b[39m.\u001b[39mto_exception(e)\n\u001b[0;32m   1270\u001b[0m   \u001b[39melse\u001b[39;00m:\n\u001b[0;32m   1271\u001b[0m     \u001b[39mraise\u001b[39;00m\n",
      "\u001b[1;31mOperatorNotAllowedInGraphError\u001b[0m: in user code:\n\n    File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\keras\\engine\\training.py\", line 1249, in train_function  *\n        return step_function(self, iterator)\n    File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\keras\\engine\\training.py\", line 1233, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\keras\\engine\\training.py\", line 1222, in run_step  **\n        outputs = model.train_step(data)\n    File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\transformers\\modeling_tf_utils.py\", line 1637, in train_step\n        y_pred = self(x, training=True)\n    File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 70, in error_handler\n        raise e.with_traceback(filtered_tb) from None\n\n    OperatorNotAllowedInGraphError: Exception encountered when calling layer 'tf_bert_for_sequence_classification' (type TFBertForSequenceClassification).\n    \n    in user code:\n    \n        File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\transformers\\modeling_tf_utils.py\", line 1557, in run_call_with_unpacked_inputs  *\n            return func(self, **unpacked_inputs)\n        File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\transformers\\models\\bert\\modeling_tf_bert.py\", line 1569, in call  *\n            outputs = self.bert(\n        File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\keras\\utils\\traceback_utils.py\", line 70, in error_handler  **\n            raise e.with_traceback(filtered_tb) from None\n    \n        OperatorNotAllowedInGraphError: Exception encountered when calling layer 'bert' (type TFBertMainLayer).\n        \n        in user code:\n        \n            File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\transformers\\modeling_tf_utils.py\", line 1557, in run_call_with_unpacked_inputs  *\n                return func(self, **unpacked_inputs)\n            File \"d:\\ProgrammFiles\\miniconda\\envs\\cv\\lib\\site-packages\\transformers\\models\\bert\\modeling_tf_bert.py\", line 766, in call  *\n                batch_size, seq_length = input_shape\n        \n            OperatorNotAllowedInGraphError: Iterating over a symbolic `tf.Tensor` is not allowed: AutoGraph did convert this function. This might indicate you are trying to use an unsupported feature.\n        \n        \n        Call arguments received by layer 'bert' (type TFBertMainLayer):\n          • self=tf.Tensor(shape=<unknown>, dtype=int32)\n          • input_ids=None\n          • attention_mask=tf.Tensor(shape=<unknown>, dtype=int32)\n          • token_type_ids=tf.Tensor(shape=<unknown>, dtype=int32)\n          • position_ids=None\n          • head_mask=None\n          • inputs_embeds=None\n          • encoder_hidden_states=None\n          • encoder_attention_mask=None\n          • past_key_values=None\n          • use_cache=None\n          • output_attentions=False\n          • output_hidden_states=False\n          • return_dict=True\n          • training=True\n    \n    \n    Call arguments received by layer 'tf_bert_for_sequence_classification' (type TFBertForSequenceClassification):\n      • self={'input_ids': 'tf.Tensor(shape=<unknown>, dtype=int32)', 'attention_mask': 'tf.Tensor(shape=<unknown>, dtype=int32)', 'token_type_ids': 'tf.Tensor(shape=<unknown>, dtype=int32)'}\n      • input_ids=None\n      • attention_mask=None\n      • token_type_ids=None\n      • position_ids=None\n      • head_mask=None\n      • inputs_embeds=None\n      • output_attentions=None\n      • output_hidden_states=None\n      • return_dict=None\n      • labels=None\n      • training=True\n"
     ]
    }
   ],
   "source": [
    "model = TFBertForSequenceClassification.from_pretrained(\"bert-base-multilingual-cased\", num_labels=3)\n",
    "optimizer = tf.keras.optimizers.Adam(learning_rate=5e-5)\n",
    "model.compile(optimizer=optimizer, loss=model.compute_loss, metrics=[\"accuracy\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.fit(train_data_tf.shuffle(1000).batch(16), epochs=3, batch_size=16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "validation_InputExamples = convert_data_to_examples_without_labels(test_data, DATA_COLUMN)\n",
    "test_data_tf = convert_examples_to_tf_dataset(list(validation_InputExamples), tokenizer)\n",
    "\n",
    "predictions = model.predict(test_data_tf.batch(16))\n",
    "predicted_labels = tf.argmax(predictions.logits, axis=1).numpy()\n",
    "\n",
    "label_reverse_map = {'Negative': 0, 'Neutral': 1, 'Positive': 2, 'Extremely Positive': 3, 'Extremely Negative': 4}\n",
    "predicted_labels_str = [label_reverse_map[label] for label in predicted_labels]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
